ModernBERT bug fixes

warner-benjamin opened the pull request:
This PR fixes a few issues with the ModernBERT implementation and some typos in the docs.

First, on the flash attention 2 path, ModernBERT was incorrectly always returning repadded logits without gradients (see #35386). This PR will return padded logits with gradients if no labels are passed to `ModernBertForMaskedLM` or if `repad_logits_with_grad` is set to true. We don't keep the gradient when repadding after using an internal loss function by default to save memory.

Second, this PR fixes torch.compile from being automatically set on when on CPU (#35388).

Third, I added some details to the model doc strings.

Fourth, documentation is updated to capitalize BERT in ModernBERT following the other BERT models.

cc @ArthurZucker @tomaarsen

warner-benjamin committed:
bug fixes

warner-benjamin committed:
organize imports

tomaarsen commented on docs/source/en/_toctree.yml:
```diff
@@ -503,7 +503,7 @@
         title: MobileBERT
       - local: model_doc/modernbert
-        title: ModernBert
+        title: ModernBERT
```
Good call - I got carried away with the Python class naming

tomaarsen commented on src/transformers/models/modernbert/modeling_modernbert.py:
```diff
@@ -667,6 +670,13 @@ def _maybe_set_compile(self):
             self.config.reference_compile = False
 
+        if self.device.type == "cpu":
+            logger.warning_once(
```
I think we should wrap this in `if self.config.reference_compile:` just like above so users don't get warned when they never manually specified anything related to `reference_compile`.

Unless you'd rather always give this warning so people aren't confused why there's a small precision-esque difference between GPU setups?

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -892,6 +900,13 @@ def _maybe_set_compile(self):
             self.config.reference_compile = False
 
+        if self.device.type == "cpu":
+            logger.warning_once(
```
I think we should wrap this in `if self.config.reference_compile:` just like above so users don't get warned when they never manually specified anything related to `reference_compile`.

Unless you'd rather always give this warning so people aren't confused why there's a small precision-esque difference between GPU setups?

warner-benjamin replied:
That was my intention, I forgot the if statement. Fixed in ab11657.

tomaarsen replied:
Nice! Resolved

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -852,12 +858,14 @@ def _autoset_attn_implementation(
                 return cls._check_and_enable_flash_attn_2(
                     config,
-                    torch_dtype=torch_dtype,
+                    torch_dtype=torch.float16,
```
I think this is a cleaner solution to avoid the unnecessary FP32 warning than I figured was possible, nice.

warner-benjamin committed:
wrap cpu warning in reference_compile

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35404). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -141,6 +142,9 @@ class ModernBertConfig(PretrainedConfig):
             shared between devices, and 4) the model is not resized after initialization. If `True`, then the model may
             be faster in some scenarios.
+        repad_logits_with_grad (`bool`, *optional*, defaults to `False`):
+            When True, ModernBertForMaskedLM keep track of the logits' gradient when repadding for output. This only
```
```suggestion
            When True, ModernBertForMaskedLM keeps track of the logits' gradient when repadding for output. This only
```

ArthurZucker reviewed (commented):
I'll let @tomaarsen update, I think we can just check if training and labels iwthout having a new arg

tomaarsen committed:
Avoid needing repad_logits_with_grad, always repad with grads when training

I'm not 100% that the conditional with "or labels is None" makes sense though - not sure what the intention is there. Perhaps we can remove that?

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -1246,8 +1257,9 @@ def forward(
 
         if self.config._attn_implementation == "flash_attention_2":
-            with torch.no_grad():
+            with nullcontext() if self.training or labels is None else torch.no_grad():
```
@warner-benjamin I just updated the `config.repad_logits_with_grad` with `self.training`, but I worry that this conditional doesn't make sense. Could you look it over? 
I don't see when we might need grads when not training, so perhaps this can just be:
```suggestion
            with nullcontext() if self.training else torch.no_grad():
```

tomaarsen replied:
I've reversed this to avoid unnecessarily increasing memory when training in some situations.

tomaarsen committed:
Merge branch 'main' of https://github.com/huggingface/transformers into pr-35404

ArthurZucker reviewed (approved):
LGTM, thanks both!

tomaarsen committed:
Revert "Avoid needing repad_logits_with_grad, always repad with grads when training"

This reverts commit cedcb4e89bcea199a1135a0933e71f534b656239.

tomaarsen committed:
Fix grammar: keep -> keeps

tomaarsen committed:
Propagate grammar fix with modular_model_converter
